### PR TITLE
DRA: don't query claims from API server

### PR DESCRIPTION
**Pull request by bart0sh**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

There is a problem in kubelet when a) a pod gets force-deleted (= removed from API server while running) and b) node-authorizer is enabled to only grant kubelet access to resource claims linked to pods on the node (= [pending PR](https://github.com/kubernetes/kubernetes/pull/116254)):
E0713 08:50:49.984737   62838 pod_workers.go:1300] "Error syncing pod, skipping" err="failed to fetch ResourceClaim external-claim referenced by pod tester-1: [resourceclaims.resource.k8s.io](http://resourceclaims.resource.k8s.io/) \"external-claim\" is forbidden: User \"system:node:127.0.0.1\" cannot get resource \"resourceclaims\" in API group \"[resource.k8s.io](http://resource.k8s.io/)\" in the namespace \"dra-3508\": no relationship found between node '127.0.0.1' and this object" pod="dra-3508/tester-1" podUID="bca3e0a6-d586-4252-bb5d-63e1b6d1d3d0"
kubelet is trying to get the claim, but cannot do so anymore and fails. NodeUnprepareResources is never called.

This PR fixes the issue by caching claim information when PrepareResources is called and using that cache in  UnprepareResources instead of querying API server.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by TommyStarK** [comment]
I do not have the experience of Patrick and Kevin, but this looks correct to me.
/lgtm

File: pkg/kubelet/cm/dra/state/checkpoint.go [hunk 1]
```diff
@@ -34,9 +41,44 @@ type DRAManagerCheckpoint struct {
...
+	// Namespace is a claim namespace
+	Namespace string
+
+	// PodUIDs is a set of pod UIDs that reference a resource
+	PodUIDs sets.Set[string]
+
+	// CDIDevices is a map of DriverName --> CDI devices returned by the
+	// GRPC API call NodePrepareResource
+	CDIDevices map[string][]string
+}
```
Comment (klueska) [inline pkg/kubelet/cm/dra/state/checkpoint.go:77]: This should go next to the declaration of the "normal" ClaimInfo state (either move that one here or this one there).

File: pkg/kubelet/cm/dra/state/checkpoint.go [hunk 2]
```diff
@@ -63,6 +105,37 @@ func (dc *DRAManagerCheckpoint) VerifyChecksum() error {
...
+		// Calculate checksum for old checkpoint
+		object := dump.ForHash(oldcheckpoint)
+		object = strings.Replace(object, "DRAManagerCheckpointWithoutResourceHandles", "DRAManagerCheckpoint", 1)
+		object = strings.Replace(object, "ClaimInfoStateListWithoutResourceHandles", "ClaimInfoStateList", 1)
+		hash := fnv.New32a()
+		fmt.Fprintf(hash, "%v", object)
+		err = nil
+		if ck != checksum.Checksum(hash.Sum32()) {
+			err = errors.ErrCorruptCheckpoint
+		}
```
Comment (klueska) [inline pkg/kubelet/cm/dra/state/checkpoint.go:137]: Can you put this in a helper function to keep VerifyChecksum call shorter.

Comment (bart0sh) [reply pkg/kubelet/cm/dra/state/checkpoint.go]: done

File: pkg/kubelet/cm/dra/state/checkpoint.go (hunk 1 above)
Comment (bart0sh) [reply pkg/kubelet/cm/dra/state/checkpoint.go]: Thank you for the reveiw! done.

**Review by tmilox — APPROVED** [review]

**Comment by TommyStarK** [comment]
/lgtm
I guess Kevin will have another look

File: pkg/kubelet/cm/dra/state/state_checkpoint.go [hunk 3]
```diff
@@ -54,11 +55,42 @@ type ClaimInfoState struct {
...
 
+	// ResourceHandles is a list of opaque resource data for processing by a specific kubelet plugin
+	ResourceHandles []resourcev1alpha2.ResourceHandle
+
 	// CDIDevices is a map of DriverName --> CDI devices returned by the
 	// GRPC API call NodePrepareResource
 	CDIDevices map[string][]string
 }
 
+// ClaimInfoStateWithoutResourceHandles is an old implementation of the ClaimInfoState
```
Comment (klueska) [inline pkg/kubelet/cm/dra/state/state_checkpoint.go:66]: We also add the TODO to remove this with beta

File: pkg/kubelet/cm/dra/state/state_checkpoint.go [hunk 4]
```diff
@@ -54,11 +55,42 @@ type ClaimInfoState struct {
...
+	PodUIDs sets.Set[string]
+
+	// CDIDevices is a map of DriverName --> CDI devices returned by the
+	// GRPC API call NodePrepareResource
+	CDIDevices map[string][]string
+}
+
+// List of claim info to store in checkpoint
+type ClaimInfoStateListWithoutResourceHandles []ClaimInfoStateWithoutResourceHandles
+
```
Comment (klueska) [inline pkg/kubelet/cm/dra/state/state_checkpoint.go:93]: This should be declared next to `ClaimInfoStateList` -- either move that one here or this one there.

File: pkg/kubelet/cm/dra/state/state_checkpoint.go (hunk 3 above)
Comment (bart0sh) [reply pkg/kubelet/cm/dra/state/state_checkpoint.go:66]: done

**Review by klueska — COMMENTED** [review]
/approve
/lgtm

File: pkg/kubelet/cm/dra/manager.go [hunk 5]
```diff
@@ -342,26 +343,8 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
...
-
-		// Grab the allocation.resourceHandles. If there are no
-		// allocation.resourceHandles, create a single resourceHandle with no
-		// content. This will trigger processing of this claim by a single
-		// kubelet plugin whose name matches resourceClaim.Status.DriverName.
-		resourceHandles := resourceClaim.Status.Allocation.ResourceHandles
-		if len(resourceHandles) == 0 {
-			resourceHandles = make([]resourcev1alpha2.ResourceHandle, 1)
-		}
-
```
Comment (TommyStarK) [inline pkg/kubelet/cm/dra/manager.go:362]: The unit tests we introduced expect that the resource claim has been properly created and we test the case when getting the resource claim from the manager fails. I guess this needs to be updated now with your changes

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/manager_test.go#L1013-L1030

Comment (bart0sh) [reply pkg/kubelet/cm/dra/manager.go:362]: @TommyStarK Thanks for pointing out to this. I understand why "failed to fetch resource claim" test case fails, but I'm still wondering why "plugin does not exist" and "should timeout" test cases started to fail. Any ideas?

Comment (bart0sh) [reply pkg/kubelet/cm/dra/manager.go:362]: Should be fixed now.

**Pull request merged**